src/components/topo/ucc_topo.c

```diff
@@ -270,3 +283,167 @@ ucc_status_t ucc_topo_get_all_numas(ucc_topo_t *topo, ucc_sbgp_t **sbgps,
+    }
+
+    *_sbgps = sbgps;
+    *n_sbgps = nnodes;
+    return UCC_OK;
+
+error:
+    for (i = 0; i < nnodes; i++) {
+        if (sbgps[i].rank_map) {
+            ucc_free(sbgps[i].rank_map);
+        }
+    }
+    ucc_free(sbgps);
+    return status;
+}
+
+ucc_status_t ucc_topo_get_all_nodes(ucc_topo_t *topo, ucc_sbgp_t **sbgps,
+                                    int *n_sbgps)
+{
+    ucc_status_t status = UCC_OK;
+
+    if (!topo->all_nodes) {
+        status = ucc_sbgp_create_all_nodes(topo, &topo->all_nodes, &topo->n_nodes);
+    }
+
+    *sbgps = topo->all_nodes;
+    *n_sbgps = topo->n_nodes;
+
+    return status;
+}
+
+ucc_status_t ucc_topo_get_node_leaders(ucc_topo_t *topo, ucc_rank_t **node_leaders_out)
+{
+    ucc_subset_t *set = &topo->set;
+    ucc_rank_t size = ucc_subset_size(set);
+    ucc_rank_t nnodes = topo->topo->nnodes;
+    ucc_rank_t i, j;
+    ucc_rank_t *ranks_seen_per_node;
+    ucc_rank_t *per_node_leaders;
+    ucc_rank_t *node_leaders;
+
+    if (topo->node_leaders) {
+        *node_leaders_out = topo->node_leaders;
+        return UCC_OK;
+    }
+
+    ucc_assert(nnodes > 1);
+
+    /* Allocate arrays */
+    node_leaders = ucc_malloc(sizeof(ucc_rank_t) * size, "node_leaders");
+    if (!node_leaders) {
+        ucc_error("failed to allocate %zd bytes for node_leaders array",
+                  size * sizeof(ucc_rank_t));
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    ranks_seen_per_node = ucc_calloc(nnodes, sizeof(ucc_rank_t), "ranks_seen_per_node");
+    if (!ranks_seen_per_node) {
+        ucc_error("failed to allocate %zd bytes for ranks_seen_per_node array",
+                  nnodes * sizeof(ucc_rank_t));
+        ucc_free(node_leaders);
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    per_node_leaders = ucc_calloc(nnodes, sizeof(ucc_rank_t), "per_node_leaders");
+    if (!ranks_seen_per_node) {
+        ucc_error("failed to allocate %zd bytes for per_node_leaders array",
+                  nnodes * sizeof(ucc_rank_t));
+        ucc_free(node_leaders);
+        ucc_free(ranks_seen_per_node);
+        return UCC_ERR_NO_MEMORY;
+    }
+
+    /* First pass: identify node leaders */
+    for (i = 0; i < size; i++) {
+        ucc_rank_t ctx_rank = ucc_ep_map_eval(set->map, i);
+        ucc_host_id_t current_host = topo->topo->procs[ctx_rank].host_id;
+
+        /* Count ranks on this node */
+        ranks_seen_per_node[current_host]++;
+        
+        /* If this is the rank we want as leader for this node, mark it */
+        if (ranks_seen_per_node[current_host] == topo->node_leader_rank_id + 1) {
+            per_node_leaders[current_host] = i;
+        }
+    }
+
+    /* Second pass: propagate node leaders to all ranks */
+    for (i = 0; i < size; i++) {
+        ucc_rank_t ctx_rank = ucc_ep_map_eval(set->map, i);
+        ucc_host_id_t current_host = topo->topo->procs[ctx_rank].host_id;
+
+        for (j = 0; j < nnodes; j++) {
```

Comment:
Instead of this being a for loop, can it be a hash table with the key being host_id and the value being the leader number? Does ucx have something for this?